mm2src/mm2_main/src/lp_swap/maker_swap.rs

```diff
@@ -1196,91 +1195,88 @@ impl MakerSwap {
+                        // TODO: We prepared for refund but didn't finalize refund. This must be breaking something for lightning.
+                        return Ok((Some(MakerSwapCommand::ConfirmTakerPaymentSpend), vec![
+                            MakerSwapEvent::TakerPaymentSpent(tx_ident),
+                            MakerSwapEvent::TakerPaymentSpendConfirmStarted,
+                        ]));
+                    },
                 },
-            }
-        }
 
-        let other_maker_coin_htlc_pub = self.r().other_maker_coin_htlc_pub;
-        let maker_coin_swap_contract_address = self.r().data.maker_coin_swap_contract_address.clone();
-        let watcher_reward = self.r().watcher_reward;
-        let spend_result = self
-            .maker_coin
-            .send_maker_refunds_payment(RefundPaymentArgs {
-                payment_tx: &maker_payment,
-                time_lock: locktime,
-                other_pubkey: other_maker_coin_htlc_pub.as_slice(),
-                tx_type_with_secret_hash: SwapTxTypeWithSecretHash::TakerOrMakerPayment {
-                    maker_secret_hash: self.secret_hash().as_slice(),
+                // Encountered an error during swap recover.
+                Err(err) => match err {
+                    // The payment tx we want to refund isn't even on-chain. There is nothing to refund/spend.
+                    RecoverSwapError::PaymentTxNotFound => {
+                        return Ok((Some(MakerSwapCommand::Finish), vec![
+                            MakerSwapEvent::MakerPaymentRefundFailed(
+                                "MakerPayment isn't even on-chain to refund it.".into(),
+                            ),
+                        ]));
+                    },
+                    // The error is unrecoverable, retrying will not fix the issue.
+                    RecoverSwapError::Irrecoverable(e) => {
+                        return Ok((Some(MakerSwapCommand::Finish), vec![
+                            MakerSwapEvent::MakerPaymentRefundFailed(ERRL!("!maker_coin.recover_funds: {}", e).into()),
+                        ]));
+                    },
+                    // The error is temporary, retrying may fix the issue.
+                    RecoverSwapError::Temporary(e) => {
+                        error!("Error {} on recover_funds, retrying in 30 seconds", e);
+                        Timer::sleep(30.).await;
+                    },
+                    // We should wait for this many seconds and try again.
+                    RecoverSwapError::WaitAndRetry(secs) => {
+                        Timer::sleep(secs as f64).await;
+                    },
+                    // The swap will be automatically recovered after the specified locktime.
+                    RecoverSwapError::AutoRecoverableAfter(locktime) => {
+                        let maker_payment = self.r().maker_payment.clone().unwrap().tx_hex;
```

Comment:
Better use `maker_payment.as_ref()` and clone tx_hex
```rust
let maker_payment = self.r().maker_payment.as_ref().unwrap().tx_hex.clone()
```

Reply:
[5e0709d](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2280/commits/5e0709d22a683c3304019872885aeaf438feb20f)